client-spark/spark3/src/main/java/org/apache/spark/shuffle/RssShuffleManager.java

```diff
@@ -518,6 +526,37 @@ public <K, V> ShuffleWriter<K, V> getWriter(
         shuffleHandleInfo);
   }
 
+  /**
+   * Provides a task attempt id that is unique for a shuffle stage.
+   *
+   * <p>We are not using context.taskAttemptId() here as this is a monotonically increasing number
+   * that is unique across the entire Spark app which can reach very large numbers, which can
+   * practically reach LONG.MAX_VALUE. That would overflow the bits in the block id.
+   *
+   * <p>Here we use the map index or task id, appended by the attempt number per task. The map index
+   * is limited by the number of partitions of a stage. The attempt number per task is limited /
+   * configured by spark.task.maxFailures (default: 4).
+   *
+   * @return a task attempt id unique for a shuffle stage
+   */
+  @VisibleForTesting
+  protected static long getTaskAttemptId(int mapIndex, int attemptNo, int maxFailures, boolean speculationEnabled) {
```

Comment:
Do we need to consider that `mapIndex` exceed the limit?

Reply:
Passing the returned `long` as `taskAttemptId` to `ClientUtils.getBlockId` will raise an error that the `taskAttemptId` is too large for the block id bit layout.

Method `getTaskAttemptId` will not `long`-overflow because the `mapIndex` is an `int`, the `maxFailures` is an `int`. With `maxFailures = Integer.MAX_VALUE`, the value of `attemptBits` is 31 and the returned value is still a valid positive `long`.

Reply:
If you prefer to fail-fast here, I can add an assertion.

Reply:
Fixed in 3abe7ae as it is always good to fail-fast with a meaningful error message.